src/libraries/Native/Unix/System.Native/pal_random.c

```diff
@@ -20,91 +21,230 @@ Generate random bytes. The generated bytes are not cryptographically strong.
+        if (fd >= 0)
+        {
+            int expected = -1;
+
+            if (!__atomic_compare_exchange_n(&sUrandomFd, &expected, fd, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
             {
-                int expected = -1;
-                if (!__atomic_compare_exchange_n(&rand_des, &expected, fd, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
-                {
-                    // Another thread has already set the rand_des
-                    close(fd);
-                }
+                // Another thread set sUrandomFd
+                close(fd);
             }
-            else if (errno == ENOENT)
-            {                
-                sMissingDevURandom = true;
+        }
+        else
+        {
+            switch (errno) {
+            case EACCES:
+            case EPERM:
+            case ENOENT:
+                sDevUrandomIsUnusable = true;
+                // Fallthrough
+            default:
+                return false;
             }
         }
+    }
 
-        if (rand_des != -1)
+    do
+    {
+        ssize_t n = read(sUrandomFd, buffer, bufferLength);
+
+        if (n == -1)
         {
-            int32_t offset = 0;
-            do
+            if (errno == EINTR)
             {
-                ssize_t n = read(rand_des, buffer + offset , (size_t)(bufferLength - offset));
-                if (n == -1)
-                {
-                    if (errno == EINTR)
-                    {
-                        continue;
-                    }
-
-                    assert(false && "read from /dev/urandom has failed");
-                    break;
-                }
-
-                offset += n;
+                continue;
             }
-            while (offset != bufferLength);
+
+            assert(false && "read from /dev/urandom has failed");
+            return false;
         }
+
+        buffer += n;
+        bufferLength -= (size_t)n;
     }
+    while (bufferLength);
+
+    return true;
+}
+
+static bool TryAllRandomFillBufferMethods(uint8_t *buffer, size_t bufferLength)
+{
+    assert(buffer != NULL);
+
+    return PlatformRandomFillBuffer(buffer, bufferLength) ||
+           PlatformRandomFillBufferUrandomFallback(buffer, bufferLength);
+}
 
-    if (!sInitializedMRand)
+static bool NeedsXorWithLrand(uint8_t *buffer, size_t bufferLength)
+{
+    // In some circumstances, data filled by
+    // TryAllRandomFillBufferMethods() will be either all zeroed out or
+    // will have all bits set.  This happens in some rare occasions in
+    // some systems (e.g.  resuming from suspend).  Try to detect this
+    // situation here and XOR every byte with something obtained from
+    // srand48().
+
+    if (bufferLength >= sizeof(uint32_t))
     {
-        srand48(time(NULL));
-        sInitializedMRand = true;
+        uint32_t first32bits;
+
+        memcpy(&first32bits, buffer, sizeof(first32bits));
+
+        return (first32bits == 0u) || (first32bits == 0xffffffffu);
```

Comment:
The memcpy rather than reinterpret cast is because of alignment concerns?

Is memcpy for only 4 bytes actually faster than accessing the four bytes individually?

Reply:
memcpy() with a constant length argument that fits in a register generates a MOV.  These two comparisons will be two cheap comparisons: only one zero comparison, and one comparison with a 32-bit immediate.